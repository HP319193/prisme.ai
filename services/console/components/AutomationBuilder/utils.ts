import { Schema } from '@prisme.ai/design-system';

const HAS_NO_FORM = ['all', 'conditions'];
export const instructionHasForm = (name: string, schema: Schema) => {
  return !HAS_NO_FORM.includes(name) && schema.type;
};
