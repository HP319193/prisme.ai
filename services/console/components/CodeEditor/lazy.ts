import dynamic from "next/dynamic";

export const CodeEditor = dynamic(import("./"), { ssr: false });

export default CodeEditor
