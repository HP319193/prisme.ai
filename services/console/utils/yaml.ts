import { ErrorObject } from "ajv";

export interface ValidationError {
  instancePath: string;
  keyword: string;
  message: string;
  params: ErrorObject;
  schemaPath: string;
}

export const getLines = (yaml: string) =>
  yaml.split(/\n/).map((line, index) => {
    const [key, value] = line.split(/\:\s?/);
    const [, indent = "", name] = key.match(/(\s+)(.*$)/) || [];
    return {
      line: index + 1,
      indent: indent.length / 2,
      name: name || key,
      value,
    };
  });

export const getLineNumberFromPath = (yaml: string, path: string) => {
  const lines = getLines(yaml);
  const pathParts = `${path}`.replace(/^\//, "").split(/\//);
  let line = 0;

  pathParts.forEach((part, index) => {
    const found = lines
      .filter(
        ({ indent, line: lineNumber }) => indent === index && lineNumber > line
      )
      .find(({ name }) => name === part);

    if (!found) {
      throw new Error("not found");
    }
    line = found.line;
  });

  return line;
};

export const findParameter = (
  yaml: string,
  { indent, parameter }: { indent: number; parameter: string }
) => {
  const lines = getLines(yaml);
  return lines.filter(
    ({ indent: i, name }) => indent === i && name === parameter
  );
};

export const findParent = (yaml: string, line: number) => {
  const lines = getLines(yaml);
  const { indent: currentIndent } =
    lines.find(({ line: l }) => line === l) || {};
  if (currentIndent === undefined) return null;

  let currentLine = line;
  while (currentLine > 0) {
    const found = lines.find(
      ({ line: l, indent }) => currentLine === l && indent === currentIndent - 1
    );
    currentLine--;
    if (found) return found;
  }
  return null;
};
