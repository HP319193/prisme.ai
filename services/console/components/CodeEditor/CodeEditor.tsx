import { forwardRef, useEffect, useImperativeHandle, useRef } from "react";
import AceEditor, { IAceEditorProps, IMarker } from "react-ace";
import ReactAce from "react-ace/lib/ace";
import "ace-builds/src-noconflict/ext-language_tools";
import "ace-builds/src-noconflict/ext-searchbox";
import "ace-builds/src-noconflict/snippets/javascript";
import "ace-builds/src-noconflict/theme-xcode";
import "ace-builds/src-noconflict/mode-javascript";
import "ace-builds/src-noconflict/mode-css";
import "ace-builds/src-noconflict/mode-html";
import "ace-builds/src-noconflict/mode-yaml";
import "ace-builds/src-noconflict/mode-json";

import { Ace } from "ace-builds";
import { addCustomAnnotations } from "../../utils/aceEditorTweaks";

export interface CodeEditorProps extends IAceEditorProps {
  mode: "javascript" | "css" | "html" | "yaml" | "json";
  value: string;
  onChange?: (value: string) => void;
  placeholder?: string;
  height?: string;
  width?: string;
  readOnly?: boolean;
  completers?: Ace.Completer[];
  annotations?: Ace.Annotation[];
  markers?: IMarker[];
  shortcuts?: {
    name: string;
    bindKey: {
      mac: string;
      win: string;
    };
    exec: () => void;
  }[];
  scrollTo?: [number, number];
  enableAutocompletion?: boolean;
  maxLines?: number;
}

export const CodeEditor = forwardRef<AceEditor, CodeEditorProps>(
  function CodeEditor(
    {
      value: initialValue,
      onChange,
      placeholder = "",
      mode,
      height = "auto",
      width = "auto",
      readOnly = false,
      completers,
      shortcuts,
      annotations,
      markers,
      scrollTo,
      style,
      ...props
    }: CodeEditorProps,
    ref
  ) {
    const value = useRef(initialValue);
    const aceRef = useRef<ReactAce>(null);
    useImperativeHandle(ref, () => aceRef.current!, []);
    //const theme = `solarized_${darkMode ? "dark" : "light"}`;
    const theme = `xcode`;

    useEffect(() => {
      if (!aceRef.current) return;
      addCustomAnnotations(aceRef.current.editor, annotations);
    }, [annotations]);

    useEffect(() => {
      if (!aceRef.current) return;
      const { editor } = aceRef.current;

      (completers || []).forEach((completer) =>
        editor.completers.push(completer)
      );
      // This code automatically opens autocomplete menu when typing `.` character
      editor.commands.on("afterExec", (e) => {
        if (e.command.name === "insertstring" && /^[\w.]$/.test(`${e.args}`)) {
          if (!aceRef.current) return;
          aceRef.current.editor.execCommand("startAutocomplete");
        }
      });

      return () => {
        (completers || []).forEach((completer) => {
          const pos = editor.completers.indexOf(completer);
          if (pos === -1) return;
          editor.completers.splice(pos, 1);
        });
      };
    }, [completers]);

    useEffect(() => {
      if (!aceRef.current) return;
      const { editor } = aceRef.current;
      (shortcuts || []).forEach((command) => {
        editor.commands.addCommand(command);
      });
    }, [shortcuts]);

    useEffect(() => {
      if (!aceRef.current) return;
      const { editor } = aceRef.current;
      Object.keys(editor.getSession().getMarkers()).forEach((key) => {
        editor.getSession().removeMarker(+key);
      });
      (markers || []).forEach((marker) => {
        const { className, endCol, endRow, startCol, startRow, type } = marker;
        const Range = ace.require("ace/range").Range;
        editor
          .getSession()
          .addMarker(
            new Range(startRow, startCol, endRow, endCol),
            className,
            type
          );
      });
    }, [markers]);

    return (
      <AceEditor
        ref={aceRef}
        style={{
          display: "flex",
          flex: 1,
          ...style,
        }}
        mode={mode}
        theme={theme}
        onChange={(v: string) => {
          value.current = v;
          onChange && onChange(v);
        }}
        value={value.current}
        enableBasicAutocompletion={true}
        enableLiveAutocompletion={true}
        enableSnippets={true}
        readOnly={readOnly}
        tabSize={2}
        placeholder={placeholder}
        showPrintMargin={true}
        showGutter={true}
        highlightActiveLine={true}
        height={height}
        width={width}
        annotations={annotations}
        {...props}
      />
    );
  }
);

export default CodeEditor;
