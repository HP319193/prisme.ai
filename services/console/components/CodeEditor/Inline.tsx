import * as React from 'react';

import CodeEditor, { CodeEditorProps } from './CodeEditor';

export const CodeEditorInline: React.FC<CodeEditorProps> = ({
  value,
  className,
  ...props
}) => {
  const singleLine = `${value}`.split(/\n/).length === 1;

  return (
    <React.Suspense fallback={<div />}>
      <CodeEditor
        {...props}
        style={{
          flex: 1,
          minHeight: '50px',
          border: '1px solid #E5E5E5',
          borderRadius: '0.3rem',
          padding: '10px',
          ...(singleLine ? undefined : { background: 'white' }),
          ...props.style,
        }}
        value={
          typeof value === 'string' ? value : JSON.stringify(value, null, '  ')
        }
        maxLines={Infinity}
        showGutter={!singleLine}
        className={`${value ? 'filled' : ''} ${className}`}
      />
    </React.Suspense>
  );
};

export default CodeEditorInline;
