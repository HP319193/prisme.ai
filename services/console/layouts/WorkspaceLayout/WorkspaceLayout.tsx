import { useRouter } from 'next/router';
import { FC, useCallback, useEffect, useMemo, useRef, useState } from 'react';
import { useTranslation } from 'react-i18next';
import Head from 'next/head';
import { useWorkspaces } from '../../components/WorkspacesProvider';
import workspaceContext, { Pagination, WorkspaceContext } from './context';
import { EventsByDay } from '.';
import api from '../../utils/api';
import { Event, Events, EventsFilters } from '@prisme.ai/sdk';
import Error404 from '../../views/Errors/404';
import { Layout, Loading, notification } from '@prisme.ai/design-system';
import { useUser } from '../../components/UserProvider';
import HeaderWorkspace from '../../components/HeaderWorkspace';
import Storage from '../../utils/Storage';
import WorkspaceSource from '../../views/WorkspaceSource';
import usePages from '../../components/PagesProvider/context';
import { useApps } from '../../components/AppsProvider';
import debounce from 'lodash/debounce';
import { usePrevious } from '../../utils/usePrevious';
import useLocalizedText from '../../utils/useLocalizedText';
import { Workspace } from '../../utils/api';

const PAGINATION_LIMIT = 15;

const getDate = (date: Date) =>
  new Date(date.getFullYear(), date.getMonth(), date.getDate());
const addEventToMap = (allEvents: EventsByDay, newEvent: Event<Date>) => {
  const { createdAt } = newEvent;
  const date = getDate(createdAt);
  const prevEvents = allEvents.get(+date) || new Set();
  if (Array.from(prevEvents).find(({ id }) => newEvent.id === id))
    return allEvents;
  const newEvents = [...Array.from(prevEvents), newEvent];
  newEvents.sort((a, b) => +b.createdAt - +a.createdAt);
  allEvents.set(+date, new Set(newEvents));
  return allEvents;
};

const getLatest = (events: Event<Date>[]) => {
  if (events.length === 0) return;
  events.sort((a, b) => +a.createdAt - +b.createdAt);
  return events[0].createdAt;
};

export const WorkspaceLayout: FC = ({ children }) => {
  const { user } = useUser();
  const {
    query: { id },
  } = useRouter();

  const { t } = useTranslation('workspaces');
  const { localize } = useLocalizedText();
  const { fetch, update, workspaces } = useWorkspaces();
  const [loading, setLoading] = useState<WorkspaceContext['loading']>(false);
  const lockEvents = useRef(false);
  const [workspace, setCurrentWorkspace] = useState<
    WorkspaceContext['workspace'] | null
  >();
  const [events, setEvents] = useState<WorkspaceContext['events']>('loading');
  const [socket, setSocket] = useState<Events>();
  const latest = useRef<Date | undefined | null>();
  const [readEvents, setReadEvents] = useState<WorkspaceContext['readEvents']>(
    new Set()
  );
  const [filters, setFilters] = useState<WorkspaceContext['filters']>({});
  const [pagination, setPagination] = useState<Pagination>({
    page: 0,
    limit: PAGINATION_LIMIT,
  });
  const [mountSourceComponent, setMountComponent] = useState(false);
  const [displaySourceView, setDisplaySourceView] = useState(false);
  const [sourceDisplayed, setSourceDisplayed] = useState(false);
  const [fullSidebar, setFullSidebar] = useState(false);

  const { fetchPages } = usePages();
  useEffect(() => {
    fetchPages(`${id}`);
  }, [fetchPages, id]);

  const { getAppInstances } = useApps();
  useEffect(() => {
    getAppInstances(`${id}`);
  }, [getAppInstances, id]);

  const displaySource = useCallback((v: boolean) => {
    setSourceDisplayed(v);
  }, []);

  // Manage source panel display
  useEffect(() => {
    if (sourceDisplayed) {
      setMountComponent(true);
    } else {
      setDisplaySourceView(false);
      setTimeout(() => setMountComponent(false), 200);
    }
  }, [sourceDisplayed]);

  // Init socket
  const workspaceId = useMemo(() => (workspace ? workspace.id : null), [
    workspace,
  ]);
  const initSocket = useCallback(async () => {
    if (!workspaceId) return;
    setSocket((socket) => {
      const socketAlreadyInstantiatedForId =
        socket && socket.workspaceId === workspaceId;

      if (!workspaceId || socketAlreadyInstantiatedForId) {
        return socket;
      }

      socket && socket.destroy();
      return socket;
    });
    setSocket(await api.streamEvents(workspaceId));
  }, [workspaceId]);
  useEffect(() => {
    initSocket();
  }, [initSocket]);

  const fetchNextEvents = useCallback(
    async (
      emptyEventsList: boolean = false,
      newPagination: Pagination | undefined = undefined,
      newFilters: EventsFilters | undefined = undefined
    ) => {
      if (!workspaceId) return;
      lockEvents.current = true;

      if (newPagination) {
        setPagination(newPagination);
      }
      if (newFilters) {
        setFilters(newFilters);
      }

      const fetched = await api.getEvents(workspaceId, {
        ...(newFilters || filters),
        ...(newPagination || pagination),
      });

      await setEvents((events) => {
        const baseEvents =
          events === 'loading' || emptyEventsList ? [] : events;

        const newEvents = new Map(baseEvents);
        fetched.forEach((event) => {
          addEventToMap(newEvents, event);
        });
        latest.current = getLatest(fetched) || null;
        return newEvents;
      });
      lockEvents.current = false;
    },
    [filters, pagination, workspaceId]
  );

  const updateFilters = useCallback(
    async (newFilters: EventsFilters) => {
      fetchNextEvents(true, { page: 0, limit: PAGINATION_LIMIT }, newFilters);
    },
    [fetchNextEvents]
  );

  const nextEvents = useCallback(async () => {
    if (!workspace || lockEvents.current || latest.current === null) return;
    fetchNextEvents(false, { ...pagination, page: pagination.page + 1 });
  }, [fetchNextEvents, pagination, workspace]);

  // Load events history
  const prevWorkspaceId = usePrevious(workspace && workspace.id);
  useEffect(() => {
    if (!workspace || lockEvents.current || workspace.id === prevWorkspaceId) {
      return;
    }
    latest.current = undefined;
    setEvents('loading');
    fetchNextEvents();
  }, [fetchNextEvents, nextEvents, prevWorkspaceId, workspace]);

  useEffect(() => {
    return () => {
      if (!workspace || workspace.id === prevWorkspaceId) return;
      socket && socket.destroy();
    };
  }, [prevWorkspaceId, socket, workspace]);

  const readEvent = useCallback(
    (eventId: string) => {
      if (!workspace) return;
      setReadEvents((readEvents) => {
        const newReads = new Set(readEvents);
        newReads.add(eventId);
        Storage.set(`readEvents-${workspace.id}`, Array.from(newReads));
        return newReads;
      });
    },
    [workspace]
  );

  const debouncedFetchWorkspace = useMemo(
    () =>
      debounce(() => {
        if (!workspaceId) return;
        fetch(workspaceId);
      }, 2000),
    [fetch, workspaceId]
  );

  // Listen to new events
  useEffect(() => {
    const listener = async (
      eventName: string,
      eventData: Prismeai.PrismeEvent
    ) => {
      const event = {
        ...eventData,
        createdAt: new Date(eventData.createdAt || ''),
      };

      await setEvents(
        addEventToMap(new Map(events === 'loading' ? [] : events), event)
      );

      if (workspace && eventName.startsWith('workspaces.')) {
        debouncedFetchWorkspace();
      }
    };
    const off = () => socket && socket.all(listener);
    return () => {
      off();
    };
  }, [socket, events, workspace, debouncedFetchWorkspace]);

  const [invalid, setInvalid] = useState<WorkspaceContext['invalid']>(false);
  const [newSource, setNewSource] = useState<WorkspaceContext['newSource']>();
  const [saving, setSaving] = useState(false);
  const [share, setShare] = useState<WorkspaceContext['share']>();

  const setCurrent = useRef(async (id: string) => {
    setLoading(true);
    await fetch(id);
    setLoading(false);
  });

  useEffect(() => {
    setCurrent.current(`${id}`);
  }, [id]);

  useEffect(() => {
    setCurrentWorkspace(workspaces.get(`${id}`));
  }, [id, workspaces]);

  const saveSource = useCallback(async () => {
    if (!newSource) return;
    setSaving(true);
    const newWorkspace = await update(newSource);
    setSaving(false);
    if (!newWorkspace) {
      return;
    }
    setCurrentWorkspace(newWorkspace);
    notification.success({
      message: t('expert.save.confirm'),
      placement: 'bottomRight',
    });
  }, [newSource, t, update]);

  const save = useCallback(
    async (workspace: Workspace) => {
      setSaving(true);
      const newWorkspace = await update(workspace);
      setSaving(false);
      if (!newWorkspace) {
        return;
      }
      setCurrentWorkspace(newWorkspace);
      notification.success({
        message: t('save.confirm'),
        placement: 'bottomRight',
      });
    },
    [t, update]
  );

  const getAppConfig = useCallback(
    async (appInstance: string) => {
      if (!workspace) return;
      return await api.getAppConfig(workspace.id, appInstance);
    },
    [workspace]
  );
  const saveAppConfig = useCallback(
    async (appInstance: string, config: any) => {
      if (!workspace) return;
      await api.updateAppConfig(workspace.id, appInstance, config);
    },
    [workspace]
  );

  const createAutomation: WorkspaceContext['createAutomation'] = useCallback(
    async (automation) => {
      if (!workspace) return null;
      const automationResult = await api.createAutomation(
        workspace,
        automation
      );
      const { slug, ...newAutomation } = automationResult;
      setCurrentWorkspace({
        ...workspace,
        automations: {
          ...workspace.automations,
          [slug]: newAutomation,
        },
      });
      return automationResult;
    },
    [workspace]
  );

  const updateAutomation: WorkspaceContext['updateAutomation'] = useCallback(
    async (slug, automation) => {
      if (!workspace) return null;
      const automationResult = await api.updateAutomation(
        workspace,
        slug,
        automation
      );

      const { slug: newSlug, ...newAutomation } = automationResult;
      const newAutomations = { ...workspace.automations };

      if (newSlug !== slug) {
        delete newAutomations[slug];
      }
      newAutomations[newSlug] = newAutomation;

      setCurrentWorkspace({
        ...workspace,
        automations: newAutomations,
      });

      return automationResult;
    },
    [workspace]
  );

  const deleteAutomation: WorkspaceContext['deleteAutomation'] = useCallback(
    async (slug) => {
      if (!workspace) return null;
      await api.deleteAutomation(workspace, slug);

      const newWorkspace = {
        ...workspace,
      };

      const { [slug]: removed, ...filteredAutomations } =
        workspace.automations || {};
      newWorkspace.automations = filteredAutomations;
      setCurrentWorkspace(newWorkspace);

      return removed;
    },
    [workspace]
  );

  if (!workspace || !user) {
    return (
      <div className="flex flex-1 justify-center align-center">
        <Loading />
      </div>
    );
  }

  if (!loading && workspace === null) {
    return <Error404 link="/workspaces" reason={t('404')} />;
  }

  return (
    <workspaceContext.Provider
      value={{
        displaySource,
        sourceDisplayed,
        invalid,
        setInvalid,
        saving,
        newSource,
        setNewSource,
        fullSidebar,
        setFullSidebar,

        workspace,
        loading,
        filters,
        updateFilters,
        save,
        saveSource,
        events,
        nextEvents,
        readEvents,
        readEvent,
        share,
        setShare,
        getAppConfig,
        saveAppConfig,
        createAutomation,
        updateAutomation,
        deleteAutomation,
        socket: socket,
      }}
    >
      <Head>
        <title>
          {t('workspace.title', { name: localize(workspace.name) })}
        </title>
        <meta
          name="description"
          content={t('workspace.description', {
            name: localize(workspace.name),
          })}
        />
      </Head>
      <div
        className={`
          absolute top-[75px] bottom-0 right-0 left-0
          bg-white
          flex flex-1
          transition-transform
          transition-duration-200
          transition-ease-in
          z-[11]
          ${displaySourceView ? '' : '-translate-y-full'}
        `}
      >
        {mountSourceComponent && (
          <WorkspaceSource onLoad={() => setDisplaySourceView(true)} />
        )}
      </div>
      <Layout Header={<HeaderWorkspace />}>{children}</Layout>
    </workspaceContext.Provider>
  );
};

export default WorkspaceLayout;
