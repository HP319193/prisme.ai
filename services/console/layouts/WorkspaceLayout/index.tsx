import { ReactElement } from 'react';
import WorkspaceLayout from './WorkspaceLayout';

export * from './context';
export * from './WorkspaceLayout';

export const getLayout = (page: ReactElement) => {
  return <WorkspaceLayout>{page}</WorkspaceLayout>;
};

export default getLayout;
