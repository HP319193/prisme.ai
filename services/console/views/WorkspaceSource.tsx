import { FC, useCallback, useEffect, useMemo, useRef, useState } from "react";
import { YAMLException } from "js-yaml";
import { useWorkspace } from "../layouts/WorkspaceLayout";
import { Workspace } from "../api/types";
import useYaml from "../utils/useYaml";
import {
  findParameter,
  findParent,
  getLineNumberFromPath,
  ValidationError,
} from "../utils/yaml";
import { generateEndpoint } from "../utils/urls";
import { useToaster } from "../layouts/Toaster";
import { useTranslation } from "next-i18next";
import { validateWorkspace } from "@prisme.ai/validation";
import CodeEditor from '../components/CodeEditor/lazy';

interface Annotation {
  row: number;
  column: number;
  text: string;
  type: "error";
}

const getEndpointAutomationName = (value: string, line: number) => {
  const { line: l = line } = findParent(`${value}`, line) || {}
  return findParent(`${value}`, l)
}

interface WorkspaceSourceProps {
  onLoad?: () => void;
}
export const WorkspaceSource: FC<WorkspaceSourceProps> = ({ onLoad }) => {
  const { t } = useTranslation("workspaces");
  const { workspace, setInvalid, setDirty, setNewSource, invalid, save } =
    useWorkspace();
  const [value, setValue] = useState<string | undefined>();
  const [annotations, setAnnotations] = useState<Annotation[]>([]);
  const { toJSON, toYaml } = useYaml();
  const ref = useRef<HTMLDivElement>(null);
  const toaster = useToaster();

  const initYaml = useCallback(async () => {
    try {
      const { id, ...json } = workspace
      const value = await toYaml(json);
      setValue(value);
    } catch (e) { }
  }, [workspace, toYaml]);
  useEffect(() => {
    initYaml();
  }, [initYaml]);

  const checkSyntaxAndReturnYAML = useCallback(
    async (value: string) => {
      if (value === undefined) return;
      try {
        setAnnotations([]);
        return { ...await toJSON<Workspace>(value), id: workspace.id };
      } catch (e) {
        const { mark, message } = e as YAMLException;
        setAnnotations([
          {
            row: mark.line,
            column: mark.position,
            text: message,
            type: "error",
          },
        ]);
      }
    },
    [toJSON, workspace.id]
  );

  const update = useCallback(
    async (newValue: string) => {
      try {
        const json = await checkSyntaxAndReturnYAML(newValue);

        if (!json) return;
        validateWorkspace(json);
        setInvalid((validateWorkspace.errors as ValidationError[]) || false);
        setNewSource(json);
        setDirty(true);
      } catch (e) {

      }
    },
    [checkSyntaxAndReturnYAML, setDirty, setInvalid, setNewSource]
  );

  useEffect(() => {
    if (!invalid || !value) {
      setAnnotations([]);
      return;
    }

    const annotations = invalid
      .map(({ instancePath, message }) => {
        try {
          const row = getLineNumberFromPath(value, instancePath) - 1;
          return {
            row,
            column: 0,
            text: message,
            type: "error",
          };
        } catch (e) { }
      })
      .filter(Boolean) as Annotation[];
    setAnnotations(annotations);
  }, [invalid, value]);

  const allAnnotations = useMemo(() => {
    if (!workspace) return annotations;
    const endpoints = findParameter(`${value}`, {
      indent: 3,
      parameter: "endpoint",
    }).map(({ line, value: v }) => ({
      row: line - 1,
      column: 0,
      text: generateEndpoint(
        workspace.id,
        v === "true" ? (getEndpointAutomationName(`${value}`, line) || { name: v }).name : v
      ),
      type: "endpoint",
    }));
    const allAnnotations = [...(annotations || []), ...endpoints];
    return allAnnotations;
  }, [annotations, value, workspace]);

  useEffect(() => {
    const { current } = ref;
    if (!current) return;
    const listener = (e: MouseEvent) => {
      const target = e.target as HTMLDivElement;
      if (!target.classList.contains("ace_endpoint")) return;
      const line = +(target.textContent || 0);
      const { text: url } =
        allAnnotations.find(
          ({ row, type }) => line - 1 === row && type === "endpoint"
        ) || {};
      if (!url) return;
      navigator.clipboard.writeText(url);
      toaster.show({
        severity: "info",
        summary: t("automations.endpoint.copied"),
      });
    };
    current.addEventListener("click", listener);

    return () => {
      current.removeEventListener("click", listener);
    };
  }, [allAnnotations, ref, t, toaster]);

  const shortcuts = useMemo(
    () => [
      {
        name: t('expert.save.help'),
        exec: save,
        bindKey: {
          mac: 'cmd-s',
          win: 'ctrl-s',
        },
      },
    ],
    [save, t],
  );

  if (value === undefined) return null;

  return (
    <div className="flex flex-1 flex-column" ref={ref}>
      <CodeEditor
        mode="yaml"
        value={value}
        onChange={update}
        annotations={allAnnotations}
        onLoad={onLoad}
        shortcuts={shortcuts}
      />
    </div>
  );
};

export default WorkspaceSource;
