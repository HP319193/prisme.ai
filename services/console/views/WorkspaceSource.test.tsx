import WorkspaceSource from "./WorkspaceSource";
import renderer, { act } from "react-test-renderer";
import { useWorkspace, WorkspaceContext } from "../layouts/WorkspaceLayout";
import useYaml from "../utils/useYaml";
import CodeEditor from "../components/CodeEditor/lazy";
import { YAMLException } from "js-yaml";
import { validateWorkspace } from "@prisme.ai/validation";
import { findParameter } from "../utils/yaml";

jest.mock("../utils/useYaml", () => {
  const toJSON = jest.fn();
  const toYaml = jest.fn();
  const useYaml = jest.fn(() => ({
    toJSON,
    toYaml,
  }));
  return useYaml;
});

jest.mock("../layouts/WorkspaceLayout", () => {
  const mock = {};
  return {
    useWorkspace: () => mock,
  };
});

jest.mock("../components/CodeEditor/lazy", () => {
  return () => null
})

jest.mock("@prisme.ai/validation", () => ({
  validateWorkspace: jest.fn()
}))

jest.mock("../utils/yaml", () => {
  const findParameter = jest.fn(() => [])
  return {
    findParameter
  }
})

it("should render empty", () => {
  const root = renderer.create(<WorkspaceSource />);
  expect(root.toJSON()).toMatchSnapshot();
});

it("should render empty", async () => {
  (useWorkspace() as WorkspaceContext).workspace = {
    id: "42",
    automations: {},
    createdAt: "2022-01-01",
    updatedAt: "2022-01-01",
    name: "foo",
  };
  (useYaml().toYaml as jest.Mock).mockImplementation(() => "foo");
  const root = renderer.create(<WorkspaceSource />);
  await act(async () => {
    await true;
  });
  expect(root.toJSON()).toMatchSnapshot();
});

it('should check syntax', async () => {
  (useWorkspace() as any).workspace = {
    name: 'foo'
  };
  (useYaml().toYaml as jest.Mock).mockImplementation(() => `
  name: foo
`)
  const root = renderer.create(<WorkspaceSource />);
  await act(async () => {
    await true;
  })
  const ed = root.root.findByType(CodeEditor)
  expect(ed.props.annotations).toEqual([]);

  (useYaml().toJSON as jest.Mock).mockImplementation(() => {
    throw new YAMLException('Invalid', {
      line: 3,
      position: 0,
      buffer: '',
      column: 0,
      snippet: 'error',
      name: 'error'
    })
  })
  act(() => {
    ed.props.onChange(`
name: foo
automations
error
`)
  })
  await act(async () => {
    await true;
  })
  expect(root.root.findByType(CodeEditor).props.annotations).toEqual([{
    row: 3,
    column: 0,
    text: `Invalid in \"error\" (4:1)

error`,
    type: 'error'
  }]);

})

it('should check workspace format', async () => {
  (useWorkspace() as any).workspace = {
    name: 'foo'
  };
  (useWorkspace() as any).setInvalid = jest.fn();
  (useYaml().toYaml as jest.Mock).mockImplementation(() => `
  name: foo
`)
  const root = renderer.create(<WorkspaceSource />);
  await act(async () => {
    await true;
  })
  const ed = root.root.findByType(CodeEditor)
  expect(ed.props.annotations).toEqual([]);

  (useYaml().toJSON as jest.Mock).mockImplementation(() => ({
    name: 'foo',
    automations: []
  }));
  (validateWorkspace as any as jest.Mock).mockImplementation(() => {
    validateWorkspace.errors = [{
      instancePath: '/automations',
      keyword: '',
      message: 'automations should be an array',
      params: {},
      schemaPath: '/automations'
    }]
    return false;
  })

  act(() => {
    ed.props.onChange(`
name: foo
automations: []
`)
  })
  await act(async () => {
    await true;
  })
  expect((useWorkspace() as any).setInvalid).toHaveBeenCalledWith([{
    instancePath: '/automations',
    keyword: '',
    message: 'automations should be an array',
    params: {},
    schemaPath: '/automations'
  }])

})

it('should find endpoints', async () => {
  (useWorkspace() as any).workspace = {
    name: 'foo',
    id: '42'
  };
  (findParameter as jest.Mock).mockImplementation(() => [{
    line: 3,
    indent: 2,
    name: 'foo',
    value: true
  }]);
  (useYaml().toYaml as jest.Mock).mockImplementation(() => `name: foo
automations:
  foo:
    when:
      endpoint: true
`)
  const root = renderer.create(<WorkspaceSource />);
  await act(async () => {
    await true;
  })
  const ed = root.root.findByType(CodeEditor)
  expect(ed.props.annotations).toEqual([{
    column: 0,
    row: 2,
    text: "http://localhost:3000/api/workspace/42/webhook/true",
    type: "endpoint",
  },])
})
