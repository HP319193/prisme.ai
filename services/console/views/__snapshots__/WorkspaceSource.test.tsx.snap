// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render empty 1`] = `null`;

exports[`should render empty 2`] = `
<div
  className="flex flex-1 flex-column"
/>
`;
